pkg/apis/resource/validation/validation.go

```diff
@@ -433,179 +512,244 @@ func validateNodeName(name string, fldPath *field.Path) field.ErrorList {
+	numFields := 0
+	if attribute.BoolValue != nil {
+		numFields++
+	}
+	if attribute.IntValue != nil {
+		numFields++
+	}
+	if attribute.StringValue != nil {
+		if len(*attribute.StringValue) > resource.DeviceAttributeMaxValueLength {
+			allErrs = append(allErrs, field.TooLongMaxLength(fldPath.Child("string"), *attribute.StringValue, resource.DeviceAttributeMaxValueLength))
+		}
+		numFields++
 	}
-	switch len(entries) {
+	if attribute.VersionValue != nil {
+		numFields++
+		if !semverRe.MatchString(*attribute.VersionValue) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("version"), *attribute.VersionValue, "must be a string compatible with semver.org spec 2.0.0"))
+		}
+		if len(*attribute.VersionValue) > resource.DeviceAttributeMaxValueLength {
+			allErrs = append(allErrs, field.TooLongMaxLength(fldPath.Child("version"), *attribute.VersionValue, resource.DeviceAttributeMaxValueLength))
+		}
+	}
+
+	switch numFields {
 	case 0:
-		allErrs = append(allErrs, field.Required(fldPath, "exactly one structured model field must be set"))
+		allErrs = append(allErrs, field.Required(fldPath, "exactly one value must be set"))
 	case 1:
 		// Okay.
 	default:
-		allErrs = append(allErrs, field.Invalid(fldPath, sets.List(entries), "exactly one field must be set, not several"))
+		allErrs = append(allErrs, field.Invalid(fldPath, attribute, "exactly one field must be set, not several"))
 	}
 	return allErrs
 }
 
-// ValidateResourceClaimParameters tests if a ResourceClaimParameters update is valid.
-func ValidateResourceClaimParametersUpdate(resourceClaimParameters, oldResourceClaimParameters *resource.ResourceClaimParameters) field.ErrorList {
-	allErrs := corevalidation.ValidateObjectMetaUpdate(&resourceClaimParameters.ObjectMeta, &oldResourceClaimParameters.ObjectMeta, field.NewPath("metadata"))
-	requestStored := apiequality.Semantic.DeepEqual(oldResourceClaimParameters.DriverRequests, resourceClaimParameters.DriverRequests)
-	allErrs = append(allErrs, validateResourceClaimParameters(resourceClaimParameters, requestStored)...)
-	return allErrs
+func validateQuantity(quantity apiresource.Quantity, fldPath *field.Path) field.ErrorList {
+	// Any parsed quantity is valid.
+	return nil
 }
 
-// ValidateResourceClassParameters tests if a ResourceClassParameters object is valid for creation.
-func ValidateResourceClassParameters(parameters *resource.ResourceClassParameters) field.ErrorList {
-	return validateResourceClassParameters(parameters, false)
+func validateQualifiedName(name resource.QualifiedName, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if name == "" {
+		allErrs = append(allErrs, field.Required(fldPath, "name required"))
+		return allErrs
+	}
+
+	// Naming the two parts in a field path is tricky. Treating them as a child field
+	// is not quite right, but close enough...
```

Comment:
in other places we say things like "prefix part" rather than lie about field path.  I think it's goofy to say:

```
invalid key: "spec.devices.basicDevice[0].attributes[example.com/not-a-c-ident].identifier"
```

See `IsQualifiedName()` in staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go

Reply:
The problem is that the functions which get called here to validate the parts don't support adding this "prefix part". They only accept the field path.

I've removed the extension of the field path. If an error then is about a part, giving that part as value should provide enough information about what part of the name is wrong. I'm treating the empty part special, because that would be ambiguous otherwise.